maximize-tensorflow-performance-on-cpu-considerations-and-recommendations-for-inference/maximize-tensorflow-performance-on-cpu-considerations-and-recommendations-for-inference.md

```diff
@@ -0,0 +1,287 @@
+<br>
+NUMA, or non-uniform memory access, is a memory layout design used in data center machines meant to take advantage of locality of memory in multi-socket machines with multiple memory controllers and blocks. Running on a NUMA-enabled machine brings with it, special considerations. Intel® Optimization for TensorFlow runs inference workload best when confining both the execution and memory usage to a single NUMA node. When running on a NUMA-enabled system, recommendation is to set intra_op_parallelism_threads to the numbers of local cores in each single NUMA-node.
+<br><br>
+Recommended settings: --cpunodebind=0 --membind=0
+<br><br>
+Usage (shell)
+<br>
+<pre>numactl --cpunodebind=0 --membind=0 python</pre>
+
+<details>
+<summary>Concurrent Execution</summary>
+<br>
+You can optimize performance by breaking up your workload into multiple data shards and then running them concurrently on more than one NUMA node. On each node (N), run the following command:
+<br><br>
+Usage (shell)
+<br>
+<pre>numactl --cpunodebind=N --membind=N python</pre>
+  For example, you can use the “&” command to launch simultaneous processes on multiple NUMA nodes:
+  <br>
+    <pre>numactl --cpunodebind=0 --membind=0 python & numactl --cpunodebind=1 --membind=1 python</pre>
+  <br>
+</details>
+
+<details>
+<summary>CPU Affinity</summary>
+  <br>
+  Users could bind threads to specific CPUs via "--physcpubind=cpus" or "-C cpus"
+  <br><br>
+  Setting its value to "0-N" will bind  threads to physical cores 0 to N only.
+  <br><br>
+  Usage (shell)
+  <pre>numactl --cpunodebind=N --membind=N -C 0-N python</pre>
+  For example, you can use the “&” command to launch simultaneous processes on multiple NUMA nodes on physical CPU 0 to 3 and 4 to 7:
+  <pre>numactl --cpunodebind=0 --membind=0 -C 0-3 python & numactl --cpunodebind=1 --membind=1 -C 4-7 python</pre>
+  NOTE : oneDNN will <a href="https://github.com/oneapi-src/oneDNN/blob/e535ef2f8cbfbee4d385153befe508c6b054305e/src/cpu/platform.cpp#LL238">get the CPU affinity mask</a> from users' numactl setting and set the maximum number of working threads in the threadpool accordingly after TensorFlow v2.5.0 RC1.
+</details>
+
+## OpenMP Technical Performance Considerations for Intel® Optimization for TensorFlow
+> This section is only for Intel® Optimization for TensorFlow, and it does not apply to official TensorFlow release.  
+<br>
+Intel® Optimization for TensorFlow utilizes OpenMP to parallelize deep learnng model execution among CPU cores.
+<br><br>
+Users can use the following environment variables to be able to tune Intel® optimized TensorFlow performance . Thus, changing values of these environment variables affects performance of the framework. These environment variables will be described in detail in the following sections. We highly recommend users tuning these values for their specific neural network model and platform.
+<br><br>
+<li>KMP_AFFINITY</li>  
```

Comment:
maybe we could also remove line 172-175 because the collapsible items also cover them

Reply:
Removed